Scrapy Project.

ansabgillani opened the pull request:
Missing Module: 

Link Extractor from HTML pages.

ansabgillani committed:
Adding Scrapy Project

DawoudSheraz commented on docket_scrap/middlewares.py:
```diff
@@ -0,0 +1,103 @@
+        spider.logger.info('Spider opened: %s' % spider.name)
+
+
+class DocketScrapDownloaderMiddleware:
```
What is the purpose of this middleware?

DawoudSheraz commented on docket_scrap/settings.py:
```diff
@@ -0,0 +1,88 @@
+# Obey robots.txt rules
+ROBOTSTXT_OBEY = True
+
+# Configure maximum concurrent requests performed by Scrapy (default: 16)
+#CONCURRENT_REQUESTS = 32
+
+# Configure a delay for requests for the same website (default: 0)
+# See https://docs.scrapy.org/en/latest/topics/settings.html#download-delay
+# See also autothrottle settings and docs
+#DOWNLOAD_DELAY = 3
+# The download delay setting will honor only one of:
+#CONCURRENT_REQUESTS_PER_DOMAIN = 16
+#CONCURRENT_REQUESTS_PER_IP = 16
+
+# Disable cookies (enabled by default)
+#COOKIES_ENABLED = False
+
+# Disable Telnet Console (enabled by default)
+#TELNETCONSOLE_ENABLED = False
+
+# Override the default request headers:
+#DEFAULT_REQUEST_HEADERS = {
+#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
+#   'Accept-Language': 'en',
+#}
+
+# Enable or disable spider middlewares
+# See https://docs.scrapy.org/en/latest/topics/spider-middleware.html
+#SPIDER_MIDDLEWARES = {
+#    'docket_scrap.middlewares.DocketScrapSpiderMiddleware': 543,
+#}
+
+# Enable or disable downloader middlewares
+# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
+#DOWNLOADER_MIDDLEWARES = {
+#    'docket_scrap.middlewares.DocketScrapDownloaderMiddleware': 543,
+#}
```
why have you commented out these settings? If they are not needed, do not commit them altogether.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+        Handling the initial response of the webpage
+        """
+
+        # Extracting the menu body of the webpage
+        body = response.css("table")[2]
+
+        # Extracting the table of the dockets
+        docket_table = body.css("table")[1]
+
+        # Importing all rows
+        rows = docket_table.xpath('tr')
+
+        # temporary storage for dockets for extraction purposes
+        temporary_storage = dict()
+
+        # Looping through each row for data
```
1. Do not add unnecessary new lines
2. You should not add in-line comments unless necessary/context is not clear without the comments. The variable names should convey the meaning properly. e.g. temporary_storage does not indicate what items will be stored in it.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+        """
+
+        # Extracting the menu body of the webpage
+        body = response.css("table")[2]
+
+        # Extracting the table of the dockets
+        docket_table = body.css("table")[1]
+
+        # Importing all rows
+        rows = docket_table.xpath('tr')
```
these 3 lines can be combined into one by using appropriate css selectors(nth-child among one)

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+            # converting a row into a list of data
+            col = row.css('td')
+
+            # if a row has three elements
+            if len(col) == 3:
```
Explain why/when can a row contain 3 elements.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+                # storing it all in temporary storage
+                temporary_storage = docket
+
+            # if the row has two elements
+            elif len(col) == 2:
```
again, explain why row will have 2 columns.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+                    # extracting file_url that resides in docket_id anchor tag
+                    file_url = self.extract_anchor(docket_id)
+                    docket["file_url"] = file_url
+
+                    # extracting date from the description
+                    date = self.extract_filing_date(description)
+                    docket["date"] = date
+
+                else:
+                    # if id is not present, then it is merged with previous id
+                    filer = col[0]
+                    filer = self.extract_filer(filer)
+                    docket["filer"] = filer
+
+                    # extracting description from second column
+                    description = col[1]
+                    description = self.extract_description(description)
+                    docket["description"] = description
+
+                    # extracting date from description
+                    date = self.extract_filing_date(description)
+                    docket["date"] = date
+
+                    # copying file_url and id from temporary storage
+                    docket["file_url"] = temporary_storage["file_url"]
+                    docket["docket_id"] = temporary_storage["docket_id"]
+
+            # Since the row length is 1, id and description are merged with previous
+            else:
+
+                # extracting filer from the column
+                filer = col[0]
+                filer = self.extract_filer(filer)
+                docket["filer"] = filer
+
+                # copying previous docket data to current
+                docket["file_url"] = temporary_storage["file_url"]
+                docket["docket_id"] = temporary_storage["docket_id"]
+                docket["description"] = temporary_storage["description"]
+                docket["date"] = temporary_storage["date"]
```
These if/else conditional blocks can be improved significantly. Only certain items are changed in if/else. The extraction calls are not varied. Move the operations that will always happen outside the conditionals. Only assign the values in if conditions.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+            yield docket
+
+    def write_in_file(self, docket):
+        # opening file for csv
+        file = open('docket.csv', 'a', encoding='UTF-8', newline='')
+
+        csv_columns = ['docket_id', 'description', 'date', 'filer', 'file_url']
+
+        # creating a csv writer for the file
+        writer = csv.DictWriter(file, csv_columns)
+
+        # generating a row
+        row = dict(docket)
+
+        # writing on csv file
+        writer.writerow(row)
```
- Use docstrings, not inline comments
- remove unnecessary newlines

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+                link = urljoin(base=self.base_url, url=link)
+            return link
+        #               Request(url=link, callback=self.fetch_pdf_links)
+        else:
+            return None
```
In python, if nothing is returned, None is returned by default. You don't need this else condition.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+    # extractor = LinkExtractor(allow='^.*\.(pdf)$')
+    # extractor.extract_links()
+
+    def extract_filing_date(self, description):
+        if description:
+            date = re.search("([0-9]{1,2}/[0-9]{1,2}/[0-9]{2,4})", description)
+            if date:
+                date = date.group()
+                return date
```
check id 4892. One of its dates is not extracted. The same has happened with some other entries. You might want to verify what is missing in your regex.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,213 @@
+        description = re.sub(' +', ' ', description)
+        if description:
+            return description
+        else:
+            return None
+
+    def extract_filer(self, filer):
+        filer = filer.css('td').extract()
+        filer = remove_tags(filer[0])
+        if filer:
+            return filer.strip()
+        else:
+            return None
+
+    def extract_docket_id(self, docket_id):
+        docket_id = docket_id.css('td').extract()
+        docket_id = remove_tags(docket_id[0])
+        docket_id = docket_id.strip()
+        if docket_id.isnumeric():
+            return docket_id
+        else: return None
```
No need to add explicit None return.

DawoudSheraz reviewed (changes requested):
- Add appropriate docstrings
- Update code quality

ansabgillani committed:
Applied changes as requested

Made changes as comments

Changed the function name from writing to append

Changed the function name from writing to append

Updated Regex and CSV

Changed the function name from writing to append

Changed regex and updated csv file

Converted text to numbered dates

Converted text to numbered dates

Removing unwanted settings

Removed docstrings to inline comments

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,179 @@
+        """
+        rows = response.css("table")[2].css("table")[1].xpath('tr')
+
+        '''
+            temporary storage for dockets to check if there is any missing field
+        '''
+        docket_temporary_storage = dict()
+
+        for row in rows:
+            col = row.css('td')
+
+            if len(col) == 3:
+                '''
+                    when a row has three columns, it must have an ID, a Filer and a Description
+                '''
```
Ok, to avoid confusion, when I said use docstrings, it means to add the relevant information in docstring. If needed, use inline comments where needed. Don't use docstring format for inline comments.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,179 @@
+                    when a row has three columns, it must have an ID, a Filer and a Description
+                '''
+
+                def extract_from_three_columns():
+                    return self.create_docket(
+                        docket_id=self.extract_docket_id(col[0]),
+                        description=self.extract_description(col[2]),
+                        date=self.extract_filing_date(col[2]),
+                        filer=self.extract_filer(col[1]),
+                        file_url=self.extract_links(col[0])
+                    )
```
This internal function is overkill. You don't need this. Just call create docket directly.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,179 @@
+
+            elif len(col) == 2:
+
+                def extract_from_two_columns():
+                    docket_id = self.extract_docket_id(col[0])
+
+                    if docket_id:
+                        ''' if id is present, then the filer is merged with previous id '''
+                        return self.create_docket(
+                            docket_id=docket_id,
+                            description=self.extract_description(col[1]),
+                            date=self.extract_filing_date(col[1]),
+                            filer=docket_temporary_storage["filer"],
+                            file_url=self.extract_links(col[0])
+                        )
+
+                    else:
+                        ''' if id not present, then it is merged with previous id '''
+                        return self.create_docket(
+                            docket_id=docket_temporary_storage["docket_id"],
+                            description=self.extract_description(col[1]),
+                            date=self.extract_filing_date(col[1]),
+                            filer=self.extract_filer(col[0]),
+                            file_url=docket_temporary_storage["file_url"]
```
same here.no need to create function. Set docket_ud, filer, and file_url in if/else condition and then call create_docket.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,179 @@
+            # id and description are merged with previous
+            else:
+
+                def extract_from_single_column():
```
remove this unneeded function as well.

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,179 @@
+
+            yield docket
+
+    def create_docket(self,
+                      docket_id,
+                      description,
+                      date,
+                      filer,
+                      file_url
+                      ):
```
👍🏽

DawoudSheraz commented on docket_scrap/spiders/docket_spider.py:
```diff
@@ -0,0 +1,179 @@
+        docket["file_url"] = file_url
+        return docket
+
+    def write_in_file(self, docket):
```
you are appending, not writing.

DawoudSheraz reviewed (commented):
added some more comments.
